morango/sync/controller.py

```diff
@@ -34,10 +43,182 @@ def deserialize_from_store(self, skip_erroring=False, filter=None):
+        from morango.sync.syncsession import NetworkSyncConnection
+
         return NetworkSyncConnection(base_url=base_url)
 
     def create_disk_connection(path):
         raise NotImplementedError("Coming soon...")
+
+
+class SessionController(object):
+    """
+    Controller class that is used to execute transfer operations, like queuing and serializing,
+    but does so through the middleware registry, through which allows customization of how
+    those transfer stage operations are handled
+    """
+
+    __slots__ = ("middleware", "context", "logging_enabled", "last_error")
+
+    def __init__(self, middleware, context=None, enable_logging=False):
+        """
+        :type middleware: morango.registry.SessionMiddlewareRegistry|list
+        :type context: morango.sync.context.SessionContext|None
+        :type enable_logging: bool
+        """
+        self.middleware = middleware
+        self.context = context
+        self.logging_enabled = enable_logging
+        self.last_error = None
+
+    @classmethod
+    def build(cls, context=None, enable_logging=False):
+        """
+        Factory method that instantiates the `SessionController` with the specified context and
+        the global middleware registry `session_middleware`
+
+        :type context: morango.sync.context.SessionContext|None
+        :type enable_logging: bool
+        :return: A new transfer controller
+        :rtype: SessionController
+        """
+        return SessionController(
+            session_middleware, context=context, enable_logging=enable_logging
+        )
+
+    def proceed_to(self, stage, context=None):
+        """
+        Calls middleware that operates on stages between the current stage and the `to_stage`. The
+        middleware are called incrementally, but in order to proceed to the next stage, the
+        middleware must return a complete status. If the middleware does not return a complete
+        status, that status is returned indicating that the method call has not reached `to_stage`.
+        Therefore middleware can perform operations asynchronously and this can be repeatedly called
+        to move forward through the transfer stages and their operations
+
+        When invoking the middleware, if the status result is:
+            PENDING: The controller will continue to invoke the middleware again when this method
+                is called repeatedly, until the status result changes
+            STARTED: The controller will not invoke any middleware until the the status changes,
+                and assumes the "started" operation will update the state itself
+            COMPLETED: The controller will proceed to invoke the middleware for the next stage
+            ERRORED: The controller will not invoke any middleware until the the status changes,
+                which would require codified resolution of the error outside of the controller
+
+        :param stage: transfer_stage.* - The transfer stage to proceed to
```

Comment:
When reading through the function, had a thought that naming this parameter `target_stage` could help with readability in some of the conditionals, but not a strong preference.